### PR TITLE
feat: XYNE-10935 open shared links in desktop app

### PR DESCRIPTION
**Pull request by XyneSpaces**
## What & why
Shared Xyne Spaces links (from Slack, pasted URLs, etc.) are plain `https` URLs, so the OS opens them in the browser. The desktop (Electron) app only registers the `xyne-spaces://` custom scheme, so nothing routed those https links to the installed app. This brings the web experience in line with mobile: when a user lands on the web app from an external context, it offers to re-open the same path in the desktop app.

## Changes (dashboard-only)
- **`apps/dashboard/src/utils/deepLink.ts`** (new): shared `DEEP_LINK_PROTOCOL`, `isElectronApp()`, `buildDeepLinkUrl()` (pathname + search only — the hash is intentionally dropped because the Electron `isSafeDeepLinkPath` guard rejects `#`), and `attemptAppLaunch()` with focus-loss detection (never redirects the browser; web page stays as fallback).
- **`apps/dashboard/src/components/OpenInAppGate/OpenInAppGate.tsx`** (new): dismissible "Open in desktop app?" banner. Shows only on genuine external entries and skips: already in Electron, iframes, auth/launch routes, reload/back-forward navigations, same-origin referrer, and once-per-tab. Choice (`always` / `never`) persisted in `localStorage` under `xyne.openInAppPreference`. Never auto-hands-off without a stored `always` preference.
- **`apps/dashboard/src/App.tsx`**: mount the gate at root.
- **`apps/dashboard/src/routes/LaunchScreen/LaunchScreen.tsx`**: consume the shared `DEEP_LINK_PROTOCOL` (removes duplicated derivation).

## Deliberately NOT changed
- No Electron changes: the generic-navigation handler in `apps/electron/src/services/deep-links.ts` already `show()` + `focus()`es the window on an incoming deep link.
- `DEEP_LINK_ROUTE_ALLOWLIST` left empty: routes are `/:workspaceId/...`, so the first path segment is a workspace cuid — populating the allowlist with route names would wrongly reject every real link.
- No backend/schema changes. Existing auth/callback, invite, and ask-ai deep-link families untouched.

## Proof of testing (POT)
- **Typecheck**: `pnpm --filter xyne-spaces-dashboard typecheck` passes clean.
- **Logic assertions** (Node): generated scheme URLs (e.g. `xyne-spaces://<wsid>/chat/<ch>/<cv>?focusThread=1`) pass a faithful port of the Electron `isSafeDeepLinkPath` guard; a hash-bearing path is rejected — proving why the hash is dropped. All 5 assertions pass.
- **Browser (dashboard dev server)**:
  - Banner appears on an external deep-link entry (empty referrer, navigation type `navigate`), captured at mount even before the app redirects to `/auth`.
  - "Don't ask again" persists `xyne.openInAppPreference=never` and removes the banner.
  - Re-entering the same link with `never` stored suppresses the banner.

## Known limitation
Exact scroll-to-message (`#origin=…&messageId=…`) is not carried into the desktop app because the Electron path guard rejects `#`; the channel/thread ids live in the path, so the app still opens the correct conversation. Carrying the hash would require coordinated Electron + renderer changes (deferred).

### CODE REVIEW

**Review by XyneSpaces — COMMENTED** [review]
Xyne Spaces Reviewer — inline findings on the specifics below; a full summary with the findings table and reviewer recommendation follows as a top-level comment. No blocking issues; these are targeted fix-ups.

File: apps/dashboard/src/components/OpenInAppGate/OpenInAppGate.tsx [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+    cleanupRef.current = attemptAppLaunch(deepLinkUrl);
+    setShowBanner(false);
+  };
+
+  const openAlways = (): void => {
+    try {
+      localStorage.setItem(PREF_KEY, 'always');
+    } catch {
+      // ignore
+    }
```
Comment (XyneSpaces) [inline apps/dashboard/src/components/OpenInAppGate/OpenInAppGate.tsx:149]: **[major] R2** — `openAlways` fires the external-protocol launch **twice**: here (manual `attemptAppLaunch`), and again when `setPreference('always')` re-triggers the effect, whose `preference === 'always'` branch calls `attemptAppLaunch(deepLinkUrl)` a second time. Two `window.location.href = xyne-spaces://…` navigations back-to-back → risk of a double OS/browser "Open app?" prompt (Chrome usually suppresses the second; Firefox/Safari may not), and the cleanup for the first timer/listener is silently overwritten in `cleanupRef`. Suggested fix: make the effect the single owner of launching — in `openAlways`, just persist `localStorage`, `markDismissedThisTab()`, `setPreference('always')`, `setShowBanner(false)` and let the effect run once.

File: apps/dashboard/src/utils/deepLink.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  let lostFocus = false;
+  const onBlur = (): void => {
+    lostFocus = true;
+  };
+  window.addEventListener('blur', onBlur);
+
+  // Triggering the scheme on a hidden anchor avoids replacing the current
+  // document if the OS has no handler (some browsers navigate on location.href).
+  try {
+    window.location.href = url;
```
Comment (XyneSpaces) [inline apps/dashboard/src/utils/deepLink.ts:61]: **[nit] R8** — comment/code mismatch: the comment above says "Triggering the scheme on a hidden anchor avoids replacing the current document", but the code sets `window.location.href` directly — no hidden anchor involved (setting `location.href` to an unknown scheme does NOT navigate the page in modern browsers, so the behavior is fine; the comment just describes a different mechanism). Also line 6: "Mirrors … and the derivation in routes/LaunchScreen/LaunchScreen.tsx — keep the **three** in step" — this PR deletes the LaunchScreen derivation, so there are now only two sources (electron `config.ts` + this file). Suggest fixing both comments to match the code.

File: apps/dashboard/src/components/OpenInAppGate/OpenInAppGate.tsx [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+ * https URL, so the OS opens it in the browser. The desktop app only registers
+ * the xyne-spaces:// custom scheme, so nothing routes an https link to it. This
+ * gate closes that gap entirely on the web side: when someone lands on the web
+ * app from an external context, it offers to re-open the same path in the desktop
+ * app via the custom scheme, and can remember that choice.
+ *
+ * It NEVER auto-hands-off without a stored "always" preference, never runs inside
+ * the desktop app itself, and always leaves the user on the web page as the
+ * fallback — a user without the app installed is unaffected.
+ */
```
Comment (XyneSpaces) [inline apps/dashboard/src/components/OpenInAppGate/OpenInAppGate.tsx:17]: **[minor] R8 / guidelines#249-261** — the `localStorage`/`sessionStorage` keys (`xyne.openInAppPreference`, `xyne.openInApp.dismissed`) are declared inline in the component. The review guidelines ask for storage keys to live in the centralized settings/constants file (or be exported from the owning module) so other consumers share one source of truth. Assumption: I haven't verified whether the dashboard already has such a constants module — if it does, please move these there; if it doesn't, exporting them from `utils/deepLink.ts` would at least keep the two keys together.

**Comment by XyneSpaces** [comment]
## Xyne Spaces Reviewer — PR #1452

**Verdict:** Changes requested — no blockers; 3 major fix-ups (1 is an Electron-side gap verified on `main`, not in this diff)
**Reviewed:** 4 files, every hunk · **Findings:** 0 blocking, 3 major, 2 minor/nit
**Scope check:** dashboard-only; no migrations touched → migration-safety gate not triggered.

### Findings
| Severity | Rule | Location | Issue |
|----------|------|----------|-------|
| major | R2 | `apps/dashboard/src/components/OpenInAppGate/OpenInAppGate.tsx:149` | "Always" fires the scheme launch twice — `openAlways` calls `attemptAppLaunch` directly, then `setPreference('always')` re-runs the effect whose `always` branch launches again → possible double OS/browser prompt; first timer's cleanup is overwritten. Fix: let the effect own the launch (persist pref + set state only). |
| major | R7 / guidelines#204 | PR-wide | New `deepLink.ts` utilities + the gate component ship with **no committed tests** — POT assertions were run ad-hoc in Node but aren't in the repo. Add unit tests (at minimum: `buildDeepLinkUrl` path/search/hash cases, `isElectronApp`, eligibility matrix). |
| major | R2 (integration; verified on `main`, **outside this diff**) | `apps/electron/src/services/deep-links.ts` — generic-nav case | Verified fact: unlike `ask-ai`/`invite` which poll up to 10s for a window, the generic branch drops the deep link when `mainWindow` is null (`if (mainWindow) {...}` after validation). A cold-start `open-url` (this PR's primary flow: user clicks "Open in app" with the app NOT running) can therefore be validated then silently dropped — app opens, right conversation never navigates to. Suggest a follow-up: wait-for-window parity in the generic branch. |
| minor | R8 / guidelines#249-261 | `OpenInAppGate.tsx:17-18` | `localStorage`/`sessionStorage` keys inline instead of a centralized settings-constants module (assumption — didn't verify whether the dashboard has one; worst case export them from `utils/deepLink.ts`). |
| nit | R8 | `deepLink.ts:61` (+ doc block at :5-6) | Comment says "hidden anchor" but code sets `window.location.href`; also "keep the three in step" — the LaunchScreen derivation was deleted by this PR, so only two sources remain. |

**Deliberate-design checks that passed:** protocol constants match `apps/electron/src/app/config.ts` for all 3 flavors ✓; `isSafeDeepLinkPath` does reject `#`, so dropping the hash is correct ✓; generic handler does `show()`+`focus()` when the window exists ✓; `window.electronAPI` is typed in `apps/dashboard/src/types/electron.d.ts` ✓; `/svgs/xyne.svg` exists ✓; eligibility gates (Electron self-skip, iframe, reload/back-forward, same-origin referrer, auth paths) all look sane.

### Recommended reviewer
**@vinit717 (Vinit Khandal)** — recommended. ⚠️ I could **not** post a formal review request: this harness has no request-reviewers capability (`gh pr edit 1452 --add-reviewer vinit717` does it manually). Tagging here instead.

| Candidate | Owns (files in this PR) | Default reviewer? | Pending reviews | Tie-breaker |
|-----------|--------------------------|-------------------|-----------------|-------------|
| **@vinit717** | None of the 4 files directly (2 are new; commits-by-path not computable with available tooling) — but authored the adjacent deep-link PR #1448 (shared links → open-in-app canonical host) | No (`.github/CODEOWNERS` covers only backend/shared ACL paths — **no entry exists for** `apps/dashboard/**`) | 0 — *no open PR in the repo currently has any requested reviewers*, so load can't differentiate | **Last seen in Spaces today (3 Sep 2026)** — most recently active |
| @ameer030 | Adjacent: merged PR #1425 (SDLC deep links) | No | 0 (see above) | Last seen ~2 months ago in Spaces |
| @samitbarai | No deep-link work found; active dashboard-area approver (#1445, #1188, #1394) | No | 0 | Last seen ~2 weeks ago |

Rationale: with no CODEOWNER eligibility set for these paths and uniform review workload, code familiarity decides — @vinit717 is actively working in exactly this feature family (shared https links → custom-scheme deep links), and is the candidate most recently active in Spaces. (Author @anuragdvd excluded; bot accounts excluded.)
